reascripts/ReaSpeech/tests/TestToolWindow.lua

```diff
@@ -0,0 +1,246 @@
+
+  local o = test_class.new()
+  lu.assertNotNil(o._tool_window)
+  lu.assertEquals(o._tool_window.is_modal, true)
+  lu.assertEquals(type(o._tool_window), 'table')
+  lu.assertEquals(o._tool_window.title, 'Test Tool Window')
+  lu.assertEquals(o._tool_window.window_flags, 21)
+  lu.assertEquals(o._tool_window.width, 300)
+  lu.assertEquals(o._tool_window.height, 200)
+end
+
+function TestToolWindow:testIsOpen()
+  local test_class = Polo {
+    init = function(self)
+      ToolWindow.init(self)
+    end
+  }
+
+  local o = test_class.new()
+  lu.assertNotNil(o._tool_window)
+  lu.assertEquals(o:is_open(), false)
+  o:open()
+  lu.assertEquals(o:is_open(), true)
+  o:close()
+  lu.assertEquals(o:is_open(), false)
+end
+
+function TestToolWindow:testOpen()
+  local test_class = Polo {
+    init = function(self)
+      ToolWindow.init(self)
+    end
+  }
+
+  ImGui = ImGui or {}
+  ImGui.WindowFlags_AlwaysAutoResize = function() return 0 end
+  ImGui.WindowFlags_NoCollapse = function() return 1 end
+  ImGui.WindowFlags_NoDocking = function() return 2 end
+
+  local o = test_class.new()
+  lu.assertNotNil(o._tool_window)
+  lu.assertEquals(o:is_open(), false)
+  o:open()
+  lu.assertEquals(o:is_open(), true)
+end
+
+function TestToolWindow:testWrapOpen()
+  local test_class = Polo {
+    init = function(self)
+      ToolWindow.init(self)
+    end
+  }
+
+  local wrapped_open_called = false
+  function test_class:open()
+    wrapped_open_called = true
+  end
+
+  local o = test_class.new()
+  o:open()
+  lu.assertEquals(o:is_open(), true)
+  lu.assertEquals(wrapped_open_called, true)
+end
+
+function TestToolWindow:testWrapClose()
+  local test_class = Polo {
+    init = function(self)
+      ToolWindow.init(self)
+    end
+  }
+
+  local wrapped_close_called = false
+  function test_class:close()
+    wrapped_close_called = true
+  end
+
+  local o = test_class.new()
+  o:open()
+  o:close()
+  lu.assertEquals(o:is_open(), false)
+  lu.assertEquals(wrapped_close_called, true)
+end
+
+function TestToolWindow:testClose()
+  local test_class = Polo {
+    init = function(self)
+      ToolWindow.init(self)
+    end
+  }
+
+  ImGui = ImGui or {}
+  ImGui.WindowFlags_AlwaysAutoResize = function() return 0 end
+  ImGui.WindowFlags_NoCollapse = function() return 1 end
+  ImGui.WindowFlags_NoDocking = function() return 2 end
+
+  local o = test_class.new()
+  lu.assertNotNil(o._tool_window)
+  lu.assertEquals(o:is_open(), false)
+  o:open()
+  lu.assertEquals(o:is_open(), true)
+  o:close()
+  lu.assertEquals(o:is_open(), false)
+end
+
+app = {
+  trap = function(self, f) return xpcall(f, function(e) print(tostring(e)) end) end
+}
+
+function TestToolWindow:testRender()
+  ImGui = ImGui or {}
+  local old_imgui = ImGui
+
+  ImGui = {}
+
+  -- visible, open
+  ImGui.Begin = function(_, _, _, _) return true, true end
+
+  ImGui.Cond_Appearing = function() return 1 end
+  ImGui.Cond_FirstUseEver = function() return 2 end
+  ImGui.End = function() end
```

Comment:
Things like this make me wish we had a mocking library. It's good to get this code under test, though. One thing that's worked for me in the past is to move ImGui-specific calls to their own helper methods, which can be redefined in tests.

Reply:
yeah, i could see doing a big pass to clean up the mocks. the code all started to look like matrix glyphs after awhile. it was (is) really hard to feel confident in what i was testing.

Reply:
What I mean is, mock_reaper also defines some of these functions like `ImGui.WindowFlags*`, so it seems redundant to also have to define them in tests. And it seems arbitrary why some would be defined in mock_reaper. With a mocking library, we'd potentially get rid of mock_reaper entirely (or make it much more bare-bones). Anyway, not a big deal at the moment.